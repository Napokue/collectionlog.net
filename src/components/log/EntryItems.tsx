import EntryItem from './EntryItem';

interface EntryItemsProps {
  name: string;
	items: CollectionLogItem[];
	killCount: CollectionLogKillCount[];
}

const EntryItems = (props: EntryItemsProps) => {
  const { name, items, killCount } = props;

  const obtainedCount = items.filter((item) => item.obtained).length;
  const itemCount = items.length;

  let obtainedClass = 'text-yellow';
  if (obtainedCount == itemCount) {
    obtainedClass = 'text-green';
  } else if (!obtainedCount) {
    obtainedClass = 'text-red';
  }

  return (
    <div id='entry-items' className='flex md:flex flex-col w-full md:w-3/4'>
      <div className='mx-2 border-b border-b-lighter shadow-log'>
        <h3 className='text-xl font-bold text-orange'>{name}</h3>
        <p className='text-orange'>Obtained: <span className={obtainedClass}>{obtainedCount}/{itemCount}</span></p>
        {killCount.length > 0 && killCount.map((kc, i) => {
          return (
            <p key={`${i}-${kc.name}`} className='text-orange'>
              {kc.name}: <span className='text-white'>{kc.amount}</span>
            </p>
          );
        })}
      </div>
      <div className='flex flex-wrap grow content-start px-2 pt-3 mb-3 overflow-y-auto shadow-log'>
        {items.length > 0 && items.map((item, i) => <EntryItem key={`${i}-${item.id}`} item={item}/>)}
      </div>
    </div>
  );
};

export default EntryItems;