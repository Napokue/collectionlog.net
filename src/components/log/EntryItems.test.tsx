import EntryItems from './EntryItems';
import { TestUtilities } from '../../tests';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

const entryName = 'Master Treasure Trails (Rare)';
const killCount: CollectionLogKillCount = {
  name: 'Master clues completed',
  amount: 1,
};

const completedCollection = Array<CollectionLogItem>(
  TestUtilities.createCollectionLogItem('3rd age longsword'),
  TestUtilities.createCollectionLogItem('3rd age wand'),
  TestUtilities.createCollectionLogItem('3rd age bow')
);

const incompleteCollection = Array<CollectionLogItem>(
  TestUtilities.createCollectionLogItem('3rd age longsword'),
  TestUtilities.createCollectionLogItem('3rd age wand'),
  TestUtilities.createCollectionLogItem('3rd age bow', false)
);

const emptyCollection = Array<CollectionLogItem>(
  TestUtilities.createCollectionLogItem('3rd age longsword', false),
  TestUtilities.createCollectionLogItem('3rd age wand', false),
  TestUtilities.createCollectionLogItem('3rd age bow', false)
);

describe(`${EntryItems.name} component`, () => {
  test.each([
    ['text-green', completedCollection],
    ['text-yellow', incompleteCollection],
    ['text-red', emptyCollection],
  ])(
    'should have obtainedClass: %p',
    (expectedObtainedClass: string, items: CollectionLogItem[]) => {
      const { container } = render(<EntryItems name={entryName} items={items} killCount={[killCount]}/>);
      const span = container.querySelector('p.text-orange span');
      expect(span).toBeTruthy();
      expect(span?.className).toBe(expectedObtainedClass);
    }
  );

  test.each([
    ['3/3', completedCollection],
    ['2/3', incompleteCollection],
    ['0/3', emptyCollection],
  ])(
    'should have completed: %p',
    (expectedCompletion: string, items: CollectionLogItem[]) => {
      const { container } = render(<EntryItems name={entryName} items={items} killCount={[killCount]}/>);
      const span = container.querySelector('p.text-orange span');
      expect(span).toBeTruthy();
      expect(span?.innerHTML).toBe(expectedCompletion);
    }
  );

  test(`should have '${killCount.name}' to be '${killCount.amount}'`, () => {
    const render = renderer.create(<EntryItems name={entryName} items={[]} killCount={[killCount]}/>);
    expect(render).toMatchSnapshot();
  });

  test('should display collection',
    () => {
      const render = renderer.create(<EntryItems name={entryName} items={completedCollection} killCount={[killCount]}/>);
      expect(render).toMatchSnapshot();
    }
  );
});