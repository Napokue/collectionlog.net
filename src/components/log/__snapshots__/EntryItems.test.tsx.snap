// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryItems component should display collection 1`] = `
<div
  className="flex md:flex flex-col w-full md:w-3/4"
  id="entry-items"
>
  <div
    className="mx-2 border-b border-b-lighter shadow-log"
  >
    <h3
      className="text-xl font-bold text-orange"
    >
      Master Treasure Trails (Rare)
    </h3>
    <p
      className="text-orange"
    >
      Obtained: 
      <span
        className="text-green"
      >
        3
        /
        3
      </span>
    </p>
    <p
      className="text-orange"
    >
      Master clues completed
      : 
      <span
        className="text-white"
      >
        1
      </span>
    </p>
  </div>
  <div
    className="flex flex-wrap grow content-start px-2 pt-3 mb-3 overflow-y-auto shadow-log"
  >
    <div
      className="w-1/4 sm:w-1/5 md:w-[16.6%] h-[50px] mb-[10px]"
    >
      <span
        className="relative top-[25%] left-[2%] md:top-[20%] md:left-[15%] text-yellow"
      >
        1
      </span>
      <div
        className="flex justify-center h-[45px] group"
      >
        <a
          className="hidden flex-col items-center justify-center mb-6 h-full group-hover:flex no-underline"
          href="https://oldschool.runescape.wiki/w/3rd_age_longsword"
          rel="noreferrer"
          target="_blank"
        >
          <p
            className="whitespace-nowrap"
          >
            3rd age longsword
          </p>
          <p>
            1/1/2023
          </p>
        </a>
        <img
          className="group-hover:hidden"
          src="data:image/jpeg;charset=utf-8;base64,undefined"
        />
      </div>
    </div>
    <div
      className="w-1/4 sm:w-1/5 md:w-[16.6%] h-[50px] mb-[10px]"
    >
      <span
        className="relative top-[25%] left-[2%] md:top-[20%] md:left-[15%] text-yellow"
      >
        1
      </span>
      <div
        className="flex justify-center h-[45px] group"
      >
        <a
          className="hidden flex-col items-center justify-center mb-6 h-full group-hover:flex no-underline"
          href="https://oldschool.runescape.wiki/w/3rd_age_wand"
          rel="noreferrer"
          target="_blank"
        >
          <p
            className="whitespace-nowrap"
          >
            3rd age wand
          </p>
          <p>
            1/1/2023
          </p>
        </a>
        <img
          className="group-hover:hidden"
          src="data:image/jpeg;charset=utf-8;base64,undefined"
        />
      </div>
    </div>
    <div
      className="w-1/4 sm:w-1/5 md:w-[16.6%] h-[50px] mb-[10px]"
    >
      <span
        className="relative top-[25%] left-[2%] md:top-[20%] md:left-[15%] text-yellow"
      >
        1
      </span>
      <div
        className="flex justify-center h-[45px] group"
      >
        <a
          className="hidden flex-col items-center justify-center mb-6 h-full group-hover:flex no-underline"
          href="https://oldschool.runescape.wiki/w/3rd_age_bow"
          rel="noreferrer"
          target="_blank"
        >
          <p
            className="whitespace-nowrap"
          >
            3rd age bow
          </p>
          <p>
            1/1/2023
          </p>
        </a>
        <img
          className="group-hover:hidden"
          src="data:image/jpeg;charset=utf-8;base64,undefined"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`EntryItems component should have 'Master clues completed' to be '1' 1`] = `
<div
  className="flex md:flex flex-col w-full md:w-3/4"
  id="entry-items"
>
  <div
    className="mx-2 border-b border-b-lighter shadow-log"
  >
    <h3
      className="text-xl font-bold text-orange"
    >
      Master Treasure Trails (Rare)
    </h3>
    <p
      className="text-orange"
    >
      Obtained: 
      <span
        className="text-green"
      >
        0
        /
        0
      </span>
    </p>
    <p
      className="text-orange"
    >
      Master clues completed
      : 
      <span
        className="text-white"
      >
        1
      </span>
    </p>
  </div>
  <div
    className="flex flex-wrap grow content-start px-2 pt-3 mb-3 overflow-y-auto shadow-log"
  />
</div>
`;
